Replaced the old config system with a new one

YouKnow-sys opened the pull request.

YouKnow-sys committed:
created the base of the new config system

YouKnow-sys committed:
fully replaced old config module with the new one
updated rest of the code to use the new config system

YouKnow-sys committed:
added `is_group_allowed` to the config

YouKnow-sys committed:
added `unauthorized_group` to `MessageText`
currently it'll always show the default english text.

YouKnow-sys committed:
update the Config file with a complete example

YouKnow-sys committed:
refactored and cleaned up the code a little

YouKnow-sys committed:
seprated the captcha to its own submodule

YouKnow-sys committed:
moved the `bot_token` to BotConfig
    now its possible to leave getting the `bot_token` to `Config` crate.
made Question `Copy`

YouKnow-sys committed:
added a command for getting bot repository link

YouKnow-sys committed:
added format check to CI

YouKnow-sys committed:
fixed admin check in the `join_check` module

hamidrezakp commented on config.toml:
```diff
@@ -1,14 +1,28 @@
-messages.wrong_answer = "❌ جواب انتخابی شما اشتباه بود، شما از گروه بن میشوید"
-messages.admin_approved_user = "✅ شما این کاربر را تایید کردید"
-messages.correct_answer = "✅ جواب شما درست بود! حالا میتوانید در گروه چت کنید"
+# list of allowed groups, if used bot will work only in these groups
```
why do we need this? in section bellow, we list all the groups and their id's and bot can only react to messages in that groups.

YouKnow-sys replied:
the reason is maybe some user want to enable the bot usage only for some group but doesn't want to modify the settings.
in other word default settings work ok for them.

hamidrezakp replied:
If we want this functionality, then the current design of settings does not fit our needs.
It should be optional for all the setting but required for group id

YouKnow-sys replied:
both settings and allowed groups Id will be optional in the next commit, user can use one without needing the other one...
user can set the allowed groups to `None` so all groups are allowed by default and then just add some custom settings to `groups_settings` if needed, or the other way around.
what exactly should change?

hamidrezakp replied:
For now it's good enough.

YouKnow-sys replied:
ok I'll go ahead and merge this to master then...

hamidrezakp commented on src/config.rs:
```diff
@@ -1,69 +1,131 @@
+#[derive(Default, Serialize, Deserialize)]
+pub struct GroupsConfig {
+    /// List of allowed group, if `None` bot will allow all groups
+    pub allowed_groups: Option<Vec<ChatId>>,
```
Why `Option` ?
isn't an empty `Vec` a sign that this option is not given and all chat's are allowed?

hamidrezakp commented on src/config.rs:
```diff
@@ -1,69 +1,131 @@
+    /// List of allowed group, if `None` bot will allow all groups
+    pub allowed_groups: Option<Vec<ChatId>>,
+    #[serde(skip)]
+    fallback_group_settings: GroupSettings,
```
I don't think we should have fallback settings. 
IMO the bot **should** only react to messages in groups that are explicitly listed.

We don't wanna to use our server CPU and RAM for some random crypto group with +1000 new users daily.

YouKnow-sys replied:
Yes, in our case, this is true. However, it is important to note that this is an open-source project. Some individuals may prefer to host the bot for their own use and may not want any limitations on the number of groups the bot can join.

hamidrezakp commented on src/config.rs:
```diff
@@ -1,69 +1,131 @@
+pub struct GroupSettings {
+    pub custom_chat_name: Option<String>,
+    /// List of allowed admin to use bot commands and admin related stuff
+    pub custom_admins: Option<Vec<UserId>>,
```
Again  here we don't need Option wrapper,
Vec[] => No custom admin
Vec[..] => Some custom admin

YouKnow-sys replied:
In this case, it was intentional. If the Vector inside the Option is empty, it will mean that admins don't have any rights to work with the bot (such as approving users and so on). This may not seem useful, but it's always good to have options.

hamidrezakp replied:
So we need an enum:
either `AllAdmin`
or `Custom(Vec<Admins>)`

YouKnow-sys replied:
hmm, yeah maybe we should use a custom enum.
although isn't using an option is good enough?
I mean `None` mean that we don't want to use a custom list for the admins
but `Some` mean we do want to use a custom list, empty or not...

hamidrezakp commented on src/join_check/captcha/math_question.rs:
```diff
@@ -38,15 +36,15 @@ impl Display for Operators {
-pub struct Question {
+#[derive(Clone, Copy)]
+pub struct MathQuestion {
     num1: u8,
```
better to be named `lhs` : Left hand side

hamidrezakp commented on src/join_check/captcha/math_question.rs:
```diff
@@ -38,15 +36,15 @@ impl Display for Operators {
+pub struct MathQuestion {
     num1: u8,
     operator: Operators,
     num2: u8,
```
and  `rhs` : Right hand side

hamidrezakp commented on src/join_check/captcha/math_question.rs:
```diff
@@ -38,15 +36,15 @@ impl Display for Operators {
-impl Question {
-    pub fn generate_question() -> (Self, [Answer; 4]) {
+impl MathQuestion {
+    pub fn generate_question() -> (Self, [u8; 4]) {
```
Not a fan of Type Aliasing? 😄 
Let me say you that by using `Answer` everyone can at one glance see that you are producing one question and 4 answers. but without it, it will be a little harder to understand at call site.

YouKnow-sys replied:
No I actually do like type aliasing 😄, but the thing is this was a internal detail and I though maybe it'll be better to just use the type itself...
but yeah, maybe it's better to use the Answer alias for  further clarification.

hamidrezakp commented on src/join_check/mod.rs:
```diff
@@ -1,55 +1,53 @@
 ) -> HandlerResult {
-    let Some(chat_config) = app_config.allowed_chats.iter().find(|c| c.id == msg.chat.id) else {
+    if !config.is_group_allowed(&msg.chat.id) {
         log::error!("Chat not found: {:?}", msg.chat);
```
Let's just keep `chat.id` and `chat.name` so we can find any misuse of our bot.
the whole struct is a little big.

YouKnow-sys replied:
Im not sure what you mean by this, can you please explain a little?
you mean like we should destruct the struct and just keep what we need?

hamidrezakp replied:
```suggestion
        log::error!("Unknown chat {} with id {}", msg.chat.title, msg.chat.id);
```

YouKnow-sys committed:
applied the comments suggestions

YouKnow-sys committed:
forgot to format the code

YouKnow-sys committed:
Update src/join_check/mod.rs

Co-authored-by: Kaveh <hamidrkp@riseup.net>

YouKnow-sys committed:
fixed the wrong formating of log message in recent commit
removed serde Serialize from all types in config
made both `allowed_groups` and `groups_settings` optional in deserializing

YouKnow-sys committed:
forgot to format again...
I'll be better if I turn on autofmt :|

hamidrezakp reviewed (approved):
LGTM
